package net.greenjab.nekomasfixed.registry.block.enums;

import net.greenjab.nekomasfixed.registry.block.HollowLogBlock;
import net.greenjab.nekomasfixed.registry.registries.BlockRegistry;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.PillarBlock;

import java.util.HashMap;
import java.util.Map;


public enum HollowLogType {

    OAK(Blocks.OAK_LOG, BlockRegistry.HOLLOW_OAK_LOG),
    SPRUCE(Blocks.SPRUCE_LOG, BlockRegistry.HOLLOW_SPRUCE_LOG),
    BIRCH(Blocks.BIRCH_LOG, BlockRegistry.HOLLOW_BIRCH_LOG),
    JUNGLE(Blocks.JUNGLE_LOG, BlockRegistry.HOLLOW_JUNGLE_LOG),
    ACACIA(Blocks.ACACIA_LOG, BlockRegistry.HOLLOW_ACACIA_LOG),
    DARK_OAK(Blocks.DARK_OAK_LOG, BlockRegistry.HOLLOW_DARK_OAK_LOG),
    MANGROVE(Blocks.CHERRY_LOG, BlockRegistry.HOLLOW_MANGROVE_LOG),
    CHERRY(Blocks.CHERRY_LOG, BlockRegistry.HOLLOW_CHERRY_LOG),
    PALE_OAK(Blocks.PALE_OAK_LOG, BlockRegistry.HOLLOW_PALE_OAK_LOG),
    BAMBOO(Blocks.PALE_OAK_LOG, BlockRegistry.HOLLOW_BAMBOO_BLOCK),
    CRIMSON(Blocks.CRIMSON_HYPHAE, BlockRegistry.HOLLOW_CRIMSON_STEM),
    WARPED(Blocks.WARPED_HYPHAE, BlockRegistry.HOLLOW_WARPED_STEM),
    BAOBAB(BlockRegistry.BAOBAB_LOG, BlockRegistry.HOLLOW_BAOBAB_LOG);

    private final Block baseLog;
    private final Block hollowLog;

    private static final Map<Block, Block> BASE_TO_HOLLOW = new HashMap<>();

    static {
        for (HollowLogType type : values()) {
            BASE_TO_HOLLOW.put(type.baseLog, type.hollowLog);
        }
    }

    HollowLogType(Block baseLog, Block hollowLog) {
        this.baseLog = baseLog;
        this.hollowLog = hollowLog;
    }

    public static Block getHollowBlock(Block baseLog) {
        return BASE_TO_HOLLOW.getOrDefault(baseLog, Blocks.AIR);
    }

    public static BlockState getHollowState(BlockState baseLog) {
        BlockState hollowState = getHollowBlock(baseLog.getBlock()).getDefaultState();
        if (!hollowState.isOf(Blocks.AIR) && hollowState.contains(PillarBlock.AXIS)) {
            return hollowState.with(HollowLogBlock.AXIS, baseLog.get(PillarBlock.AXIS));
        }
        return hollowState;
    }
}